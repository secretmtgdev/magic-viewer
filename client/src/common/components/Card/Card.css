.card-container {
    display: inline-block;
    margin: 1.4em;
}
