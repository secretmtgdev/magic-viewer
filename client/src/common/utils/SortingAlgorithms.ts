import { TCard, sortingType } from "./Types";

export const bubbleSort = (cards: TCard[], sortType: sortingType): TCard[] => {
    if(cards.length < 2) return cards;

    for(let i = cards.length; i > 0; i--) {
        let swapped: boolean = false;
        for(let j = 1; j < i; j++) {
            let firstCMC = getSortValueByType(cards[j - 1], sortType);
            let secondCMC = getSortValueByType(cards[j], sortType);
            if(firstCMC > secondCMC) {
                swap(cards, j, j - 1);
                swapped = true;
            }
        }

        if(!swapped) {
            break;
        }
    }

    return cards;
}

export const selectionSort = (cards: TCard[], sortType: sortingType): TCard[] => {
    let n: number = cards.length;
    for(let i = 0; i < n - 1; i++) {
        let minIdx = i;
        let minVal = getSortValueByType(cards[minIdx], sortType);
        for(let j = i + 1; j < n; j++) {
            let valHere = getSortValueByType(cards[j], sortType);
            if (valHere < minVal) {
                minIdx = j;
                minVal = valHere;
            }
        }

        swap(cards, minIdx, i);
    }

    return cards;
}

export const insertionSort = (cards: TCard[], sortType: sortingType): TCard[] => {
    let n: number = cards.length;
    for(let i = 1; i < n; i++) {
        let key = cards[i];
        let j = i - 1;
        while(j >= 0 && getSortValueByType(cards[j], sortType) > getSortValueByType(key, sortType)) {
            cards[j + 1] = cards[j];
            j -= 1;
        }

        cards[j + 1] = key;
    }

    return cards;
}

export const mergeSort = (cards: TCard[], sortType: sortingType): TCard[] => {
    return mergeSortHelper(cards, sortType, 0, cards.length);
}

const mergeSortHelper = (cards: TCard[], sortType: sortingType, start: number, end: number): TCard[] => {
    if(start >= end) {
        return cards;
    }

    const midPoint = start + Math.floor((end - start) / 2);
    let leftSort = mergeSortHelper(cards, sortType, start, midPoint);
    let rightSort = mergeSortHelper(cards, sortType, midPoint + 1, end);
    return merge(leftSort, rightSort, sortType, start, midPoint, end);
}

const merge = (cardsA: TCard[], cardsB: TCard[], sortType: sortingType, aStart: number, midPoint: number, bEnd: number): TCard[] => {
    let aLen = midPoint - aStart + 1;
    let bLen = bEnd - midPoint;

    let tempA = new Array(aLen).fill(0);
    let tempB = new Array(bLen).fill(0);
    let cardsC = new Array(aLen + bLen).fill(0);

    for(let i = 0; i < aLen; i++) tempA[i] = cardsA[i];
    for(let i = 0; i < bLen; i++) tempB[i] = cardsB[i];

    console.error(cardsA);
    console.error(cardsB);
    console.error(tempA);
    console.error(tempB);

    let i = 0;
    let j = 0;
    let k = 0;

    while (i < aLen && j < bLen) {
        let aVal = getSortValueByType(cardsA[i], sortType);
        let bVal = getSortValueByType(cardsB[j], sortType);
        if (tempA[i] <= tempB[j]) {
            cardsC[k] = tempA[i];
            i++;
        }
        else {
            cardsC[k] = tempB[j];
            j++;
        }
        k++;
    }
 
    // Copy the remaining elements of
    // L[], if there are any
    while (i < aLen) {
        cardsC[k] = tempA[i];
        i++;
        k++;
    }
 
    // Copy the remaining elements of
    // R[], if there are any
    while (j < bLen) {
        cardsC[k] = tempB[j];
        j++;
        k++;
    }

    return cardsC;
}

/********************
 ** HELPER METHODS ** 
 ********************/
const getSortValueByType = (card: TCard, sortType: sortingType) => {
    return sortType === 'cmc' ? card.manaValue : 
        sortType === 'price' ? convertCurrencyToNumber(card.price.nonFoil) : -1;
}

const swap = (arr: TCard[], i: number, j: number) => {
    let tmp = arr[i];
    arr[i] = arr[j];
    arr[j] = tmp;
}

const convertCurrencyToNumber = (currency: string) => {
    currency = currency.replace(/[^0-9.-]+/g, '');
    return parseFloat(currency);
}
