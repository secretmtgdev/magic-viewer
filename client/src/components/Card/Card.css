.cardContainer {
    display: inline-block;
}

.cardContainer:hover {
    cursor: pointer;
}

.cardContainer img:hover {
    border: 10px solid blue;
    transition: 350ms ease;
    border-radius: 10px;
}
