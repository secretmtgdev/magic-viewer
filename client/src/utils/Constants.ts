import { TCurrencyInfo } from "./Types"

export const CardConstants = {
    DEFAULT_CARD_HEIGHT: 275,
    DEFAULT_CARD_WIDTH: 200
}

export const TitleConstants = {
    DEFAULT_WORDS_IN_PHRASE: 3,
    DEFAULT_LETTERS_IN_WORD: 5,
    DEFAULT_LETTERS_IN_PHRASE: 15
}

export const Currency: {[key: string]: TCurrencyInfo} = {
    'en': {
        abbreviation: 'usd',
        name: 'English',
        symbol: '$'
    },
    'eur': {
        abbreviation: 'eur',
        name: 'European',
        symbol: 'â‚¬'
    }
}

export const Default = {
    Decklist: {
        name: '',
        commander: '',
        sections: []
    } 
}

export const TestMocks = {
    CardMock: {
        card: {
            name: 'card',
            manaValue: 1,
            price: {
                currency: 'usd',
                nonFoil: '1.0',
                foil: '2.0'
            },
            imgUrl: 'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcT_YCalozv20Oki516pDRvDiNfByLm_OxvEZc6_XN6TIw&s'
        }
    }
}