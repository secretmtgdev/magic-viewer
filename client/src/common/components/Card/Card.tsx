import React from 'react';

import CardCallout from '../CardCallout/CardCallout';
import CardImage from '../CardImage/CardImage';
import PriceContainer from '../PriceContainer/PriceContainer';

import { useState } from 'react';
import { ICardProps } from '../../utils/Interfaces';

import './Card.css';

// Component name must start with a capital letter
export default function Card(cardProps: ICardProps) {
    const [hiddenState, setHiddenState] = useState(true);
    const card = cardProps.card;

    // Multiline jsx must be wrapped in parenthesis
    return (
        <>
            <div                  
                onClick={() => setHiddenState(!hiddenState)}
                id={card.name}
                className='card-container'
                >
                <CardImage {...cardProps} />
                <div>Name: {card.name}</div>
                <PriceContainer {...card.price} />
            </div>
            <CardCallout cardProps={cardProps} hidden={hiddenState} />
        </>        
    );
}