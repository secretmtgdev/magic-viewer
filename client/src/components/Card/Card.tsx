import CardCallout from '../CardCallout/CardCallout';
import CardImage from '../CardImage/CardImage';
import PriceContainer from '../PriceContainer/PriceContainer';
import React from 'react';

import { useState } from 'react';
import { ICardProps } from '../../utils/Interfaces';

import './Card.css';

// Component name must start with a capital letter
export default function Card(cardProps: ICardProps) {
    const [hiddenState, setHiddenState] = useState(true);

    // Multiline jsx must be wrapped in parenthesis
    return (
        <>
            <div 
                className='cardContainer' 
                onClick={() => setHiddenState(!hiddenState)}
                id={cardProps.card.name}>
                <CardImage {...cardProps} />
                <PriceContainer {...cardProps.card.price} />
            </div>
            <CardCallout cardProps={cardProps} hidden={hiddenState} />
        </>        
    );
}