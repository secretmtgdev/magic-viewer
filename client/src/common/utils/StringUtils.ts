import { TitleConstants } from "./Constants";

export const formalizeWord = (word: string) => {
    const firstCapital = word[0].toLocaleUpperCase() + word.slice(1);
    const splitPhrase = firstCapital.split(/(?=[A-Z])/);
    return splitPhrase.join(' ');
}

export const hashPhrase = (phrase: string): number => {
    let hashcode: number = 0;
    for (let i = 0; i < phrase.length; i += 1) {
        hashcode = (hashcode * Math.pow(37, i + 1) + phrase.charCodeAt(i)) % TitleConstants.DEFAULT_LETTERS_IN_PHRASE;
    }

    return hashcode;
}
