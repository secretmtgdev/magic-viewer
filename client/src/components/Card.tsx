import { CardConstants} from '../utils/Constants';
import { ICardProps } from '../utils/Interfaces';

// Component name must start with a capital letter
export default function Card({card, size}: ICardProps) {
    // Multiline jsx must be wrapped in parenthesis
    return (
        // Only one root element may be returned because JSX is an object
        <span className='cardContainer'>
            <img
                // Curly braces enable JavaScript in markup
                src={card.imgUrl}

                // Curly braces can only be used as texxt or attributes
                alt={card.name}
                height={size ?? CardConstants.DEFAULT_CARD_HEIGHT}
                width={size ?? CardConstants.DEFAULT_CARD_WIDTH}
            />
        </span>
    );
}