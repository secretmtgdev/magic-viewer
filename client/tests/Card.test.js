import Card from '../src/components/Card/Card';
import {configure, shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import { TestMocks } from '../src/utils/Constants';

configure({adapter: new Adapter()});

it('Toggles the card callout visibility after click', () => {
    const card = shallow(<Card {...TestMocks.CardMock} />);
    console.error(card);
});