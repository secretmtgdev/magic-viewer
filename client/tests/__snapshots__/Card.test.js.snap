// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders the card 1`] = `
[
  <div
    className="cardContainer"
    id="card"
    onClick={[Function]}
  >
    <span
      className="card-image-wrapper lazy-load-image-background "
      style={
        {
          "backgroundImage": "url(../../images/placeholder.jgpeg)",
          "backgroundSize": "100% 100%",
          "color": "transparent",
          "display": "inline-block",
          "height": 275,
          "width": 200,
        }
      }
    >
      <span
        className=""
        style={
          {
            "display": "inline-block",
            "height": 275,
            "width": 200,
          }
        }
      />
    </span>
    <p>
      Currency
      : 
      usd
    </p>
    <p>
      NonFoil
      : 
      1.0
    </p>
    <p>
      Foil
      : 
      2.0
    </p>
  </div>,
  <span
    className="ms-layer"
  >
    <div
      className="ms-Fabric ms-Layer-content content-111"
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDoubleClick={[Function]}
      onDrag={[Function]}
      onDragEnd={[Function]}
      onDragEnter={[Function]}
      onDragExit={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
      onFocus={[Function]}
      onInput={[Function]}
      onKeyDown={[Function]}
      onKeyPress={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseMove={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      onMouseUp={[Function]}
      onSubmit={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
    >
      <div
        className="ms-Callout-container container-113"
        style={
          {
            "visibility": "hidden",
          }
        }
      >
        <div
          className="ms-Callout root-114"
          hidden={true}
          id="card-callout"
          style={
            {
              "filter": "opacity(0)",
              "opacity": 0,
              "pointerEvents": "none",
            }
          }
          tabIndex={-1}
        >
          <div
            className="ms-Callout-main calloutMain-117"
            onKeyDown={[Function]}
            onMouseDown={[Function]}
            onMouseUp={[Function]}
            style={
              {
                "maxHeight": "100%",
                "outline": "none",
                "overflowY": undefined,
              }
            }
          >
            <span
              className="card-image-wrapper lazy-load-image-background "
              style={
                {
                  "backgroundImage": "url(../../images/placeholder.jgpeg)",
                  "backgroundSize": "100% 100%",
                  "color": "transparent",
                  "display": "inline-block",
                  "height": 275,
                  "width": 200,
                }
              }
            >
              <span
                className=""
                style={
                  {
                    "display": "inline-block",
                    "height": 275,
                    "width": 200,
                  }
                }
              />
            </span>
          </div>
        </div>
      </div>
    </div>
  </span>,
]
`;
